### ISSUE TITLE
TRAC Constraint-Acquisition Receipt Pack for neural-oracle-to-symbolic learning

### ISSUE THREAD

**Issue by crates** [entif-ai/rosetta#1590]
## Summary

Design and implement a **TRAC Constraint-Acquisition Receipt Pack** plus a thin interoperability adapter for the neuro-symbolic workflow in [Learning Symbolic Constraint Representations from Examples: A Neuro-Symbolic Approach](https://arxiv.org/abs/2609.12267) and its [TRAC reference implementation](https://github.com/NassimBelmecheri/TRAC).

The integration should preserve the full authority chain from labeled examples and constraint vocabulary through T-Oracle membership answers, FastCA bias mutations, the learned constraint network, equivalence evaluation, and any decision to admit that network for downstream use. It must not collapse “the neural oracle answered no,” “FastCA added a constraint,” “the learned network matched the target on this benchmark,” and “this network is authorized for planning” into one claim.

This is a prospective Rosetta integration proposal. The paper authors have not endorsed it.

## Research basis

The paper defines a closed-loop neuro-symbolic system:

- A constraint-acquisition vocabulary ((X,D)), constraint language (Gamma), and finite bias (B) bound the candidate symbolic model.
- T-Oracle classifies full or partial assignments as valid or invalid.
- The TO3 variant trains on constraint-specific tuples restricted to target scopes; this alignment between training objects and acquisition queries is empirically decisive.
- FastCA exhaustively visits the finite bias, issues partial membership queries, removes candidates or invokes `FindC`, and returns an explicit learned network (L).
- Under a correct oracle and the assumption (T subseteq B), FastCA is sound, complete, terminating, and bounded by (|B|) queries.
- With a learned oracle, errors have asymmetric consequences: false positives can remove valid constraints and reduce recall, while false negatives can retain redundant constraints and reduce precision.
- The best reported pairing, FastCA/TO3, frequently reaches 100% constraint-level accuracy, recall, and precision on the tested benchmarks, but performance is weaker on some structurally difficult problems and exhaustive search may not scale to very large biases.

Relevant sources:

- Paper: https://arxiv.org/abs/2609.12267
- Full paper: https://arxiv.org/html/2609.12267
- Code: https://github.com/NassimBelmecheri/TRAC
- Entif research context: https://entif.ai/tags/research/2026/09/12/prepaying-semantics/
- Rosetta repository: https://github.com/entif-ai/rosetta/
- Rosetta Core Spine: `docs/RFCs/Rosetta v3.0.0 Core Spine Specification.md`
- Pack contract: `docs/RFCs/ROCK-3111-C-RRP-Pack-Filesystem-Contract-v0.2.0.md`

## Why this belongs in Rosetta

TRAC has a particularly clear interoperability boundary: neural judgments become inputs to symbolic state changes. The learned constraint network is interpretable, but interpretability alone does not establish the source, reliability, or permitted scope of each constraint.

Rosetta can add a replayable authority layer around that handoff:

1. preserve the exact example, scope, vocabulary, bias, checkpoint, and threshold behind each oracle answer;
2. record each query and bias mutation as a distinct event;
3. distinguish a neural verdict from a ground-truth oracle response;
4. evaluate the final network under explicit implication/equivalence rules;
5. issue a separate receipt for promotion, restriction, supersession, or revocation;
6. let a downstream solver determine whether a constraint network is admissible for a specific task.

This is an implementable Rosetta Pack and adapter, not a request to move TRAC into Rosetta or redefine Core semantics.

## Concept and primitive crosswalk

| TRAC construct | Rosetta representation | Correspondence |
|---|---|---|
| Benchmark instance and labeled assignment | `rosetta.observation` with source/config references | Exact |
| Vocabulary ((X,D)), constraint language (Gamma), bias (B) | Versioned Pack configuration artifacts | Partial; Rosetta needs a CSP-specific schema |
| T-Oracle checkpoint and sibling JSON metadata | Content-addressed artifact plus model/config Observation | Exact |
| Membership query over a scoped partial assignment | `rosetta.action` in an acquisition Run | Exact |
| T-Oracle invocation | `rosetta.toolcall` | Exact |
| Probability, dependency score, threshold, yes/no answer | Raw Observation plus classification Evaluation | Exact if the adapter can access the underlying values |
| Candidate constraint | `rosetta.conjecture` or Pack-scoped candidate artifact | Partial; must not be asserted as accepted fact |
| Remove candidate / invoke `FindC` / add learned constraint | `rosetta.action` with before/after bias references | Exact |
| Learned network (L) | Content-addressed derived artifact | Exact |
| Precision, recall, F1, implication, equivalence | `rosetta.evaluation` | Exact |
| Approval for downstream solver use | `rosetta.receipt` with scope and caveats | Exact |
| Later correction, retraining, or replacement | New artifact plus supersession/revocation receipt | Exact |

Semantic gaps that must remain explicit:

- A constraint vocabulary is not Rosetta Core vocabulary.
- T-Oracle’s probability is not a calibrated probability unless separately established.
- FastCA’s proof assumes a correct oracle; it does not transfer unchanged to a fallible neural oracle.
- Constraint-level equivalence to a benchmark target does not establish domain validity outside that benchmark.
- A high-scoring learned network does not authorize consequential use without a separate policy decision.

## Recommended collaboration artifact

Create a **SchemaPack with a reference adapter**, tentatively named `trac-constraint-acquisition`.

A Pack is preferable to a Core change because the new semantics are domain-specific and can be expressed through:

- JSON Schemas for vocabulary, bias snapshots, scoped membership queries, oracle verdicts, mutations, learned networks, and promotion receipts;
- controlled vocabulary entries for query and mutation kinds;
- deterministic translation rules from TRAC run output;
- receipt fixtures for representative TO1/TO2/TO3 episodes;
- conformance and replay tests.

The Pack should reuse Core Run/Action/ToolCall/Observation/Evaluation/Conjecture/Receipt meanings and must pass the repository’s semantic-governance checks before introducing any durable term.

## Proposed architecture

```mermaid
flowchart LR
  A[TRAC dataset and run metadata] --> B[TRAC JSONL event exporter]
  C[Checkpoint plus config digest] --> B
  D[FastCA query and mutation events] --> B
  B --> E[Constraint acquisition adapter]
  E --> F[Rosetta Core tiles]
  E --> G[Pack-scoped CSP artifacts]
  F --> H[Receipt bundle]
  G --> H
  H --> I[Conformance and replay evaluator]
  I --> J{Promotion policy}
  J -->|pass| K[Scoped reliance receipt]
  J -->|fail or uncertain| L[Restricted or no-promotion receipt]
```

Boundaries:

- TRAC remains the system that trains T-Oracle and runs FastCA.
- The adapter consumes a stable, documented export rather than importing private Python internals.
- Rosetta canonicalizes and hashes the exported objects, emits Core spine events, validates Pack schemas, and compiles a receipt bundle.
- Promotion policy is downstream of benchmark evaluation and is never inferred from an oracle answer alone.
- Missing checkpoint digests, unknown constraint-language terms, broken query-to-mutation lineage, or schema-invalid events fail closed for promotion while preserving the raw source artifact.

## Translation and interoperability design

Define a versioned event envelope with at least:

- `run_id`, `event_id`, timestamp/ordinal, event type, and parent event IDs;
- benchmark name, scale, seed, dataset digest, and source mode;
- vocabulary variables and domains;
- constraint-language identifier and serializable relation/operator form;
- immutable bias snapshot or content digest before and after each mutation;
- T-Oracle variant, checkpoint digest, configuration digest, model dimensions, and training-data digest;
- membership query assignment, variable scope, candidate constraint, and query origin;
- raw model probability, dependency score, threshold, derived yes/no verdict, and inference device/version when exposed;
- `FindC` inputs and outputs;
- learned-network artifact and ordered constraint list;
- target-network reference when evaluation uses ground truth;
- implication policy/version and precision, recall, F1, accuracy, time, and query-count observations;
- limitations, missing fields, and reliance scope.

Translation rules:

- emit the query request as Action, model invocation as ToolCall, raw numeric/model output as Observation, and the interpreted verdict as Evaluation;
- represent each candidate constraint as a Conjecture until an explicit acquisition Action changes the learned-network state;
- link every bias mutation to the exact query/verdict that caused it;
- content-address every bias and learned-network snapshot;
- issue promotion receipts only after a named evaluation policy passes;
- preserve failed, uncertain, and contradictory episodes rather than dropping them;
- never label a neural answer as `ground_truth` unless the source actually was the benchmark oracle.

## Repository and implementation impact

Proposed Rosetta paths, subject to repository maintainers’ naming review:

- `packs/trac-constraint-acquisition/pack.json`
- `packs/trac-constraint-acquisition/spec.md`
- `packs/trac-constraint-acquisition/acceptance.md`
- `packs/trac-constraint-acquisition/metadata.json`
- `packs/trac-constraint-acquisition/schema/*.schema.json`
- `packs/trac-constraint-acquisition/vocab/*.json`
- `packs/trac-constraint-acquisition/translators/trac-event-map.json`
- `packs/trac-constraint-acquisition/test-vectors/*.json`
- `packs/trac-constraint-acquisition/receipts-fixtures/*.json`
- `packages/trac-constraint-adapter/src/lib/*.ts`
- `packages/trac-constraint-adapter/src/lib/*.spec.ts`
- `docs/packs/PACK_SUITE_INDEX.md`
- `docs/governance/genesis/SEMANTIC_AUDIT.md`

Potential upstream TRAC addition, proposed to the researchers rather than assumed:

- `trac/export.py` or an equivalent callback interface that emits stable JSONL events for generation, training, prediction, query, mutation, acquisition, and evaluation.

No runtime dependency on PyTorch or the bundled `pycona` code is required for the first Rosetta slice.

### Phase 0: authority and license review

- Confirm Pack naming and reuse of Core semantics.
- Determine whether TRAC code has an explicit license.
- Freeze a minimal interchange schema from public paper/code behavior without copying implementation code.

### Phase 1: schemas and static fixtures

- Define the event envelope and CSP artifact schemas.
- Create hand-authored, paper-grounded Sudoku fixtures for positive, negative, uncertain, and lineage-broken cases.
- Add Pack manifest, acceptance criteria, and schema validation.

### Phase 2: reference adapter

- Implement JSONL ingestion and deterministic translation to Rosetta tiles.
- Hash checkpoint/config, dataset, bias, query, verdict, mutation, and network artifacts.
- Emit a receipt bundle and replay report.

### Phase 3: TRAC exporter collaboration

- Agree with the researchers on the smallest stable export surface.
- Add or document callbacks around `trac/oracle.py`, `trac/fastca.py`, `trac/acquire.py`, and checkpoint metadata.
- Validate end-to-end runs for TO1 and TO3.

### Phase 4: promotion and failure evaluation

- Implement policy profiles for benchmark-only, advisory, and solver-admissible networks.
- Test false-positive and false-negative oracle paths.
- Add supersession/revocation after checkpoint or bias changes.

### Phase 5: scale and interoperability study

- Measure receipt volume and hashing/storage cost as (|B|) grows.
- Compare snapshot, delta, and checkpointed bias encodings.
- Publish a small cross-project example and findings without claiming author endorsement.

## Detailed task checklist

- [ ] Run Rosetta semantic-governance review for all proposed terms.
- [ ] Define canonical serialization for variables, domains, scopes, and constraint relations.
- [ ] Define `acquisition-run`, `membership-query`, `oracle-verdict`, `bias-snapshot`, `bias-mutation`, `learned-network`, and `network-evaluation` schemas.
- [ ] Define raw-output versus interpreted-verdict fields.
- [ ] Define checkpoint, dataset, configuration, and code-version digests.
- [ ] Define query-to-mutation and mutation-to-network lineage.
- [ ] Define no-answer, malformed-answer, timeout, and contradictory-answer states.
- [ ] Define benchmark-target and real-domain evaluation profiles separately.
- [ ] Implement the JSONL adapter with deterministic ordering.
- [ ] Emit Core Run/Action/ToolCall/Observation/Evaluation/Conjecture/Receipt tiles.
- [ ] Add Pack conformance tests and negative fixtures.
- [ ] Add TO1 and TO3 replay fixtures demonstrating precision collapse versus scope-aligned success.
- [ ] Add false-positive and false-negative mutation fixtures.
- [ ] Add supersession when the oracle checkpoint, threshold, vocabulary, or bias changes.
- [ ] Add documentation for a minimal end-to-end replay.
- [ ] Update the Pack suite index and semantic audit.
- [ ] Ask the TRAC authors to review the interchange fields and error semantics.

## Acceptance criteria and verification

- [ ] Every accepted or removed constraint traces to a specific query, scoped assignment, oracle invocation, raw output, interpreted verdict, and bias state.
- [ ] Replaying a fixture yields byte-stable canonical artifacts and CIDs.
- [ ] A missing checkpoint/config digest prevents a solver-admissible receipt.
- [ ] An unknown relation/operator fails schema validation without discarding the source event.
- [ ] Neural verdicts cannot be serialized as ground truth without an explicit ground-truth source type.
- [ ] TO1 and TO3 fixtures preserve their distinct training-data provenance.
- [ ] False-positive and false-negative fixtures produce distinguishable downstream effects.
- [ ] Learned-network evaluation records the implication/equivalence policy and target-network reference.
- [ ] Promotion, restriction, supersession, and revocation are separate receipts.
- [ ] Pack examples validate against the schemas; malformed and lineage-broken negative fixtures fail as intended.
- [ ] The adapter has unit tests for deterministic ordering, hashing, missing parents, duplicate event IDs, and version drift.
- [ ] `pnpm run governance:semantic`, targeted package tests, typecheck, and Pack conformance checks pass.

## Risks, constraints, and licensing

- The paper is CC BY 4.0, but the public TRAC repository currently exposes no obvious license in its root listing. Do not vendor, translate, or copy TRAC implementation code until licensing is clarified.
- TO3 assumes partial knowledge of target constraint scopes; that makes it powerful but limits claims of knowledge-free discovery.
- FastCA’s proofs require (T subseteq B) and a correct oracle. Receipts must identify when those premises are assumptions rather than established facts.
- Exhaustive bias traversal trades query minimization for latency and error robustness; artifact volume may become large.
- Oracle probability and dependency score may not be calibrated.
- Integer finite-domain benchmarks do not establish transfer to richer symbolic languages.
- Constraint equivalence can depend on the representation, solver, and implication procedure.
- The adapter must avoid making Rosetta a Python/PyTorch runtime or silently granting authority to a benchmark-trained model.

## Open questions for the researchers

1. Can TRAC expose the raw probability, dependency score, learned threshold, and candidate scope for each oracle query?
2. What is the canonical serialization of constraints and logical implication in the current code?
3. Which state transitions inside FastCA are stable enough to expose as an event callback?
4. How should a no-answer or low-confidence oracle response affect the bias?
5. Does TO3’s scope information come from the target network in every benchmark, and how should that dependency be disclosed?
6. Are checkpoint JSON files sufficient to reconstruct all inference semantics, including preprocessing and constraint ordering?
7. Would the authors consider adding an explicit license and a minimal JSONL export?
8. Which benchmark best demonstrates a genuinely ambiguous or erroneous oracle decision for a public fixture?

## Non-goals

- Reimplementing T-Oracle or FastCA in TypeScript.
- Claiming Rosetta improves TRAC’s model accuracy or acquisition speed.
- Treating neural answers as formal proofs.
- Promoting benchmark-equivalent networks for consequential use by default.
- Adding CSP-specific terms to Rosetta Core.
- Copying unlicensed TRAC code or pretrained assets.
- Replacing TRAC’s current CLI, UI, Python API, or reproduction workflow.

## Collaboration path

Start with a repository-local schema and two static replay fixtures. Share the crosswalk and event envelope with the TRAC authors for correction before requesting any upstream exporter work. A useful first joint milestone is one TO3 Sudoku acquisition episode that can be replayed from labeled-example digest through final network evaluation, plus one deliberately misclassified query showing how a neural error changes the bias and downstream receipt.

The collaboration is strictly non-commercial. Rosetta and Bithkuil are free, open-source projects, not intended for sale or monetization, and intended to pursue universally applicable advances in AI that could benefit humanity if successfully realized.

## Source references

- https://arxiv.org/abs/2609.12267
- https://arxiv.org/html/2609.12267
- https://github.com/NassimBelmecheri/TRAC
- https://entif.ai/tags/research/2026/09/12/prepaying-semantics/
- https://github.com/entif-ai/bithkuil
- https://github.com/entif-ai/rosetta/
- https://github.com/entif-ai/rosetta/blob/main/docs/RFCs/Rosetta%20v3.0.0%20Core%20Spine%20Specification.md
- https://github.com/entif-ai/rosetta/blob/main/docs/RFCs/ROCK-3111-C-RRP-Pack-Filesystem-Contract-v0.2.0.md
